src/Features/CSharp/Portable/ConvertLinq/CSharpConvertForEachToLinqQueryProvider.cs

```diff
@@ -0,0 +1,585 @@
+        }
+
+        protected override bool TryBuildSpecificConverter(
+            ForEachStatementSyntax forEachStatement,
+            SemanticModel semanticModel,
+            ImmutableArray<SyntaxNode> convertingNodes,
+            StatementSyntax statementCannotBeConverted,
+            CancellationToken cancellationToken,
+            out IConverter converter)
+        {
+            switch (statementCannotBeConverted.Kind())
+            {
+                case SyntaxKind.ExpressionStatement:
+                    var expression = ((ExpressionStatementSyntax)statementCannotBeConverted).Expression;
+                    switch (expression.Kind())
+                    {
+                        case SyntaxKind.PostIncrementExpression:
+                            // No matter what can be used as the last select statement for the case of Count. We use SyntaxFactory.IdentifierName(forEachStatement.Identifier).
+                            converter = new ConvertToCountSyntaxEditor(
+                                forEachStatement,
+                                convertingNodes,
+                                SyntaxFactory.IdentifierName(forEachStatement.Identifier),
+                                ((PostfixUnaryExpressionSyntax)expression).Operand);
+                            return true;
+
+                        case SyntaxKind.InvocationExpression:
+                            var invocationExpression = (InvocationExpressionSyntax)expression;
+                            if (invocationExpression.Expression is MemberAccessExpressionSyntax memberAccessExpression &&
+                                semanticModel.GetSymbolInfo(memberAccessExpression, cancellationToken).Symbol is IMethodSymbol methodSymbol &&
+                                IsList(methodSymbol.ContainingType, semanticModel) &&
+                                methodSymbol.Name.Equals(nameof(IList.Add)) &&
+                                methodSymbol.Parameters.Length == 1)
+                            {
+                                converter = new ConvertToListSyntaxEditor(
+                                    forEachStatement,
+                                    convertingNodes,
+                                    invocationExpression.ArgumentList.Arguments.Single().Expression,
+                                    memberAccessExpression.Expression);
+                                return true;
+                            }
+
+                            break;
+                    }
+
+                    break;
+
+                case SyntaxKind.YieldReturnStatement:
+                    var memberDeclaration = semanticModel.GetEnclosingSymbol(forEachStatement.SpanStart, cancellationToken).DeclaringSyntaxReferences.Single().GetSyntax();
+                    var yieldStatements = memberDeclaration.DescendantNodes().OfType<YieldStatementSyntax>();
```

Comment:
diving arbitrarily deep concerns me.  what if the yields are in a nested local function?